src/target/ext/tracepoints.rs

```diff
@@ -0,0 +1,523 @@
+//! Provide tracepoints for the target.
+use crate::conn::Connection;
+use crate::protocol::ResponseWriter;
+use crate::protocol::ResponseWriterError;
+use crate::target::Arch;
+use crate::target::Target;
+use crate::target::TargetResult;
+use managed::ManagedSlice;
+use num_traits::PrimInt;
+
+/// A tracepoint, identified by a unique number.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct Tracepoint(pub usize);
+
+/// A state variable, identified by a unique number.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct StateVariable(usize);
+
+/// Describes a new tracepoint. It may be configured by later
+/// [DefineTracepoint] structs. GDB may ask for the state of current
+/// tracepoints, which are described with this same structure.
+#[derive(Debug, Clone)]
+pub struct NewTracepoint<U> {
+    /// The tracepoint number
+    pub number: Tracepoint,
+    /// If the tracepoint is enabled or not
+    pub enabled: bool,
+    /// The address the tracepoint is set at.
+    pub addr: U,
+    /// The tracepoint's step count
+    pub step_count: u64,
+    /// The tracepoint's pass count.
+    pub pass_count: u64,
+    /// If there will be tracepoint "define" packets that follow this.
+    pub more: bool,
+}
+
+#[cfg(feature = "alloc")]
+impl<U: Copy> NewTracepoint<U> {
+    /// Allocate an owned copy of this structure.
+    pub fn get_owned(&self) -> NewTracepoint<U> {
+        self.clone()
+    }
+}
+
+impl<U: crate::internal::BeBytes + num_traits::Zero + PrimInt> NewTracepoint<U> {
+    pub(crate) fn write<C: Connection>(
+        &self,
+        res: &mut ResponseWriter<'_, C>,
+    ) -> Result<(), ResponseWriterError<C::Error>> {
+        res.write_str("QTDP:")?;
+        res.write_num(self.number.0)?;
+        res.write_str(":")?;
+        let mut buf = [0; 8];
+        self.addr
+            .to_be_bytes(&mut buf)
+            .ok_or_else(|| unreachable!())?;
+        res.write_hex_buf(&buf)?;
+        res.write_str(":")?;
+        res.write_str(if self.enabled { "E" } else { "D" })?;
+        res.write_str(":")?;
+        res.write_num(self.step_count)?;
+        res.write_str(":")?;
+        res.write_num(self.pass_count)?;
+
+        Ok(())
+    }
+}
+
+/// Describes how to collect information for a trace frame when the tracepoint
+/// it is attached to is hit. A tracepoint may have more than one action
+/// attached.
+#[derive(Debug)]
+#[allow(missing_docs)]
+pub enum TracepointAction<'a, U> {
+    /// Collect the registers whose bits are set in `mask` (big endian).
+    /// Note that `mask` may be larger than the word length.
+    Registers { mask: ManagedSlice<'a, u8> },
```

Comment:
Right, I wasn't sure of any other good way to do this unfortunately. The registers mask may be arbitrary size, and target implementations have to be able to return a `TracepointAction` back to gdbstub for the enumeration API which kind of precludes doing borrows or callbacks without extreme pain. If you have a better idea I'm definitely open to it.

Reply:
Some food for thought:

APIs in `gdbstub` tend to lean more towards callback-based / "write-into-a-function-provided-buffer"-based mechanisms of reporting this data, as it grants the target implementation more flexibility in how it wishes to collect / store / report data back to `gdbstub`.

A prime example of this principles in play is in the [MultiThreadBase::list_active_threads](https://docs.rs/gdbstub/latest/gdbstub/target/ext/base/multithread/trait.MultiThreadBase.html#tymethod.list_active_threads) API, which _could_ have been modeled as a `fn() -> &[Tid]`, but in doing so, it would've _required_ target implementation to allocate a buffer to stash thread-ids for this operation, which could be wasteful on embedded systems. With a callback based approach, target implementations can dynamically compute what Tids are active, and report them in an "ad-hoc" manner back to `gdbstub` as part of whatever transient iteration / query it happens to be using to determine active thread-ids.

I'll note that a callback-based API - while certainly more annoying to handle within `gdbstub` - isn't actually _that_ much more annoying for end-users, especially if the callbacks are encapsulated in a `struct CallbackHelper()` type that exposes some useful methods for users to pass in arrays (if they have already have them).

In general, `gdbstub` will _always_ prioritize complicating its internals if it means the end-user API can be made more useful / ergonomic for end users (in this particular case, this would be more regarding the "more useful" aspect of the calculus, given the difference in "power-level" between returning a slice vs. using callbacks).

* * *

One other observation, separate from the analysis above: if we stick with the current design of returning "long-lived" slices / references to target-owned data, then I don't actually think we'd need to expose `managed` in the API? Presumably, we could refine the types at play to just use basic slices / primitives?